File: Python/RemoveDuplicatesFromSortedList2.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)   
        dummy.next = head
        prev = dummy
        current = head

        while current:
            while current.next and current.val == current.next.val:
                current = current.next
            if prev.next == current:
                prev = prev.next
            else:
                prev.next = current.next
            current = current.next
        return dummy.next
            
            
            
def list_to_ll(ls):
    if not ls:
        return None
    head=ListNode(ls[0])
    current=head
    for i in ls[1:]:
        current.next=ListNode(i)
        current=current.next
    return head

def print_ll(head):
    curr=head
    while curr:
        print(curr.val,end=' -> ')
        curr=curr.next
    print('None')
    
s=Solution()
head=[1,2,3,3,4,4,5]
head=list_to_ll(head)
print('Before deletion')
print_ll(head)
head=s.deleteDuplicates(head)
print('After deletion')
print_ll(head)

File: LeetCode/Java/RemoveDuplicatesFromSortedList2.java
// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class RemoveDuplicatesFromSortedList2 {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        ListNode current = head;

        while (current != null) {
            while (current.next != null && current.val == current.next.val) {
                current = current.next;
            }
            if (prev.next == current) {
                prev = prev.next;
            } else {
                prev.next = current.next;
            }
            current = current.next;
        }
        return dummy.next;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1,
                new ListNode(2,
                new ListNode(3,
                new ListNode(3,
                new ListNode(4,
                new ListNode(4,
                new ListNode(5)))))));

        RemoveDuplicatesFromSortedList2 sol = new RemoveDuplicatesFromSortedList2();
        ListNode result = sol.deleteDuplicates(head);

        while (result != null) {
            System.out.print(result.val);
            if (result.next != null) System.out.print("->");
            result = result.next;
        }
    }
}
